iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp

```diff
@@ -61,41 +67,37 @@ class function_ref;
 ///         callback();
 /// @endcode
 template <class ReturnType, class... ArgTypes>
-class function_ref<ReturnType(ArgTypes...)>
+class function_ref<ReturnType(ArgTypes...)> final
 {
-    using SignatureType = ReturnType(ArgTypes...);
-
-    template <typename T1, typename T2>
-    using has_same_decayed_type = typename std::integral_constant<
-        bool,
-        bool(std::is_same<typename std::decay<T1>::type, typename std::decay<T2>::type>::value)>;
-
   public:
     ~function_ref() noexcept = default;
 
     function_ref(const function_ref&) noexcept = default;
 
-    function_ref& operator=(const function_ref&) noexcept = default;
+    function_ref& operator=(const function_ref&) & noexcept = default;
```

Comment:
Why did you add a reference here? Was it a warning? Should we do this for all copy assignments?

So this rule forbids code like:
```cxx
std::move(a) = b;
```
Which is valid and well defined but on the first glimpse weird. But take a look at this here

```cxx
template<typename T>
void initializeSomeArgs(T&& args) {
    std::forward<T>(args) = someT;
}
```

Here we take `args` as universal reference (so sometimes also as rvalue reference) and would like to assign a new value to it. 
Why should we forbid that? What kind of error can possible originate from that?

Reply:
@elfenpiff That's due to AUTOSAR  rule `A12-8-7`. Indeed

```
std::move(a) = b;
```
is weird and does not make sense. You can find more example in the AUTOSAR document.

For your 2nd example, I suppose if it is needed in some cases, we should explicitly define `foo& operator=(const foo&) &&` and if desired call `operator=() &` from there explicitly.

Reply:
If both implementations (`&&` and `&`) exist and are identical it does not make sense to distinguish. Only if one should not exist (as here) or if they both exist but are different.